src/libponyc/pass/refer.c

```diff
@@ -66,6 +66,121 @@ static bool is_this_incomplete(pass_opt_t* opt, ast_t* ast)
   return false;
 }
 
+static const char* generate_multi_dot_name(ast_t* ast, ast_t** def_found) {
+  pony_assert(ast_id(ast) == TK_DOT);
+
+  ast_t* def = NULL;
+  size_t len = 0;
+  ast_t* temp_ast = ast;
+
+  do {
+    if(ast_id(temp_ast) != TK_DOT)
+      break;
+
+    AST_GET_CHILDREN(temp_ast, left, right);
+    def = (ast_t*) ast_data(left);
+    temp_ast = left;
+
+    // the `+ 1` is for the '.' needed in the string
+    len += strlen(ast_name(right)) + 1;
+  } while(def == NULL);
+
+  switch(ast_id(temp_ast))
+  {
+    case TK_DOT:
+    {
+      AST_GET_CHILDREN(temp_ast, left, right);
+      if(ast_id(left) == TK_THIS)
+      {
+        temp_ast = right;
+        len += strlen(ast_name(temp_ast));
+      }
+      else
+        pony_assert(0);
+      break;
+    }
+
+    case TK_LETREF:
+    case TK_VARREF:
+    case TK_REFERENCE:
+    case TK_PARAMREF:
+    {
+      temp_ast = ast_child(temp_ast);
+      len += strlen(ast_name(temp_ast));
+      break;
+    }
+
+    case TK_THIS:
+    {
+      temp_ast = ast_sibling(temp_ast);
+      // string len already added by loop above
+      // subtract 1 because we don't have to add '.'
+      // since we're ignoring the `this`
+      len -= 1;
+      break;
+    }
+
+    default:
+    {
+      pony_assert(0);
+    }
+  }
+
+  if(def_found != NULL)
+  {
+    *def_found = def;
+
+    if(def == NULL)
+      return stringtab("");
+  }
+
+  // for the \0 at the end
+  len = len + 1;
+
+  char* buf = (char*)ponyint_pool_alloc_size(len);
+  size_t offset = 0;
+  const char* name = ast_name(temp_ast);
+  size_t slen = strlen(name);
+  strncpy(buf + offset, name, slen);
+  offset += slen;
+  temp_ast = ast_parent(temp_ast);
+
+  while(temp_ast != ast)
+  {
+    buf[offset] = '.';
+    offset += 1;
+    temp_ast = ast_sibling(temp_ast);
+    name = ast_name(temp_ast);
+    slen = strlen(name);
+    strncpy(buf + offset, name, slen);
+    offset += slen;
+    temp_ast = ast_parent(temp_ast);
+  }
+
+  pony_assert((offset + 1) == len);
+  buf[offset] = '\0';
+
+  return stringtab_consume(buf, len);
+}
+
+static bool is_match(ast_t* a, ast_t* b)
```

Comment:
Maybe a more descriptive name could be used here? Not sure exactly what would be best, but perhaps `is_matching_assign_lhs`?

Reply:
okey dokey.